pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -62,11 +62,13 @@ var userRegistriesDir = filepath.FromSlash(".config/containers/registries.conf.d
 // The function errors if the newly created reference is not parsable.
 func (e *Endpoint) rewriteReference(ref reference.Named, prefix string) (reference.Named, error) {
```

Comment:
What happens on user input (prefix=*.foo, location unset)? AFAICS that’s converted into (prefix=*.foo, location=*.foo), and this rewrite function then turns `domain.foo/bar` into `*.foo/bar`, breaking the format.

(There should be a test for this as well.)

Reply:
(I have re-formatted the above to show asterisks instead of being interpreted as italic.)

Reply:
will do, thanks

Reply:
> What happens on user input (prefix=`*.foo`, location unset)? 

So, just to confirm, do you mean this:
```
# Location unset with wildcarded prefix
[[registry]]
location=""
prefix="*.foo"
```

So, unset location doesn't seem to be accepted to begin with and `make test` panics along with a:
`Error:      	Expected nil, but got: error loading registries configuration "testdata/find-registry.conf": invalid location: cannot be empty` 

I'll look at adding a test for panics if we need it in.

Reply:
err .. let me double check on this ...

Reply:
My mistake, I meant `location="*.foo"` + `prefix=""`  (or an explicit `location="*.foo"`+`prefix="*.foo"`).

The test panic is probably just something using `assert.Nil(t, err)` instead of `require.NoError(t, err)`.

Reply:
Sorry, what's the acceptable behavior in case of location containing `*.`. Should there be an additional validation check in postProcessRegistries to return an error on this format? Or, should it proceed but not actually do the rewrite ?

Reply:
My very tentative guess:
- Externally in the config file, keep allowing (location=$foo, prefix=""), but only for non-wildcards. Keep post-processing that into internal representation (location=$foo, prefix=$foo), for compatibility with existing API users.
- New, allow (prefix=$foo, location=""), both in the config file and after post-processing, meaning “match but don’t rewrite”; in that case, $foo can contain wilcards. Document this as the preferred form of user configuration; document the Prefix/Location fields as deprecated.
- Validate that Location does not contain wildcards. After post-processing, Registry.Location might be empty (but mirrors’ locations must not be).

That’s how I conceptually think about this (Prefix is for matching configuration, so must always be present; Location for redirections is optional and rarely used) — but I haven’t really examined the full scope in detail, e.g. how `postProcessRegistries` does the consistency check of insecure/blocked for `Location`.

Reply:
> What happens on user input (prefix=*.foo, location unset)? AFAICS that’s converted into (prefix=*.foo, location=*.foo), and this rewrite function then turns domain.foo/bar into *.foo/bar, breaking the format.

The `location` must always be set, so I don't think there's a problem.

Reply:
Only the prefix can empty, in which case it will default to the location's value.

Reply:
Yeah, but we can end up with both Prefix and Location containing `*.foo`, and the rewrite logic currently interprets that as inserting the `*.` into the result.

Reply:
> My mistake, I meant location="*.foo" + prefix="" (or an explicit location="*.foo"+prefix="*.foo").

Ah, right, that's a problem but it predates this PR.  We didn't do any validation of the prefixes so far and only a very minimal one the location.

While I am supportive to improve that part, I don't think it's critical to this PR. Also considering the deadline tomorrow.

Reply:
The original https://issues.redhat.com/browse/RFE-1001 IIRC asks for a feature that allows setting `*.foo` as “insecure”; i.e. a wildcard prefix, some options, _no remapping_. If we don’t allow empty Location, that’s currently impossible to configure. The user can either set `Location=*.foo`, or `Location=bar`. Either way, that doesn’t do what the user wanted, to let requests for `host.foo/…` connect to `host.foo/…` but disable TLS.

Reply:
Reading more of the RFE, it’s unclear to me whether they actually _need_ the insecure flag or whether it was incidental, and primarily then wanted the allowed/blocked registries. Either way, the “blocked registries” option shows up both in `policy.json` and `registries.conf` per https://github.com/openshift/runtime-utils/blob/master/pkg/registries/registries.go and https://github.com/openshift/machine-config-operator/blob/ca8fcb887f153604556f2706cee8a33165879797/pkg/controller/container-runtime-config/helpers.go#L381 , so I think we do need need to have wildcard matching without triggering a remapping.

---

It feels unlikely to me the primary feature that must get in before the deadline is “remap `*.foo/…` to `bar/…`”, but it’s possible; I’m afraid I can’t now review the full history of the related RFEs. Either way prioritization is not up to me.

Reply:
Your points convince me. Thanks, @mtrmac!

@lsm5.  I agree that we need to support the `location=""` case in which case no remapping will happen (i.e., the provided input will be used) _but_ the matching should work.

Reply:
```toml
[[registry]]
prefix="*.internal.registry.com
insecure=true
```

In that case, a `podman pull xxx.internal.registry.com/image:tag` would pull from `xxx. ...` without TLS (insecure).
`blocked=true` would work implicitly

Reply:
ack, I'll add a check for this and may also need code change I guess.

Reply:
@vrothberg I first added and then removed the check for `insecure=true` based on @mtrmac's later review comment and @umohnani8 said openshift/mco didn't need it. Current version only checks for empty location and wildcarded prefix.